src/Microsoft.Azure.WebJobs.Host/JobHost.cs

```diff
@@ -329,37 +335,92 @@ private static IFunctionDefinition ResolveFunctionDefinition(MethodInfo method,
             return function;
         }
 
-        private async Task<JobHostContext> CreateContextAndLogHostStartedAsync(CancellationToken cancellationToken)
+        private void ThrowIfDisposed()
         {
-            JobHostContext context = await _config.CreateAndLogHostStartedAsync(this, _shutdownTokenSource.Token, cancellationToken);
+            if (_disposed)
+            {
+                throw new ObjectDisposedException(null);
+            }
+        }
 
-            lock (_contextLock)
+        // If multiple threads call this, only one should do the init. The rest should wait
+        // When this task is signalled, _context is initialized. 
+        private Task EnsureHostStartedAsync(CancellationToken cancellationToken)
+        {
+            if (_context != null)
             {
-                if (_context == null)
+                return Task.CompletedTask;
+            }
+
+            TaskCompletionSource<bool> tsc = null;
+
+            lock (_stopTaskLock)
```

Comment:
This lock object is for the "stop" operation/function. Why are you locking on it here?